### PR TITLE
Rnn make stepnet member

### PR DESCRIPTION
**Pull request by Superjom**
fix: https://github.com/PaddlePaddle/Paddle/issues/3457

### CODE REVIEW

**Comment by Superjom** [comment]
fix: https://github.com/PaddlePaddle/Paddle/issues/3457

File: paddle/operators/recurrent_op.h [hunk 1]
```diff
@@ -117,10 +130,14 @@ class RecurrentOp final : public framework::OperatorBase {
...
   }
 
+  void set_stepnet(std::shared_ptr<NetOp> net) { stepnet_ = net; }
+  const NetOp* stepnet() const { return stepnet_.get(); }
+
   static const rnn::ArgumentName kArgName;
 
  private:
   RecurrentAlgorithm alg_;
+  std::shared_ptr<NetOp> stepnet_;
```
Comment (wangkuiyi) [inline paddle/operators/recurrent_op.h]: Why would we need to define `stepnet_` as a `std::shared_ptr`? Can we define

```cpp
class RecurrentOp ... {
 private:
  NetOp step_;
```

File: paddle/operators/recurrent_op.h [hunk 2]
```diff
@@ -33,7 +34,12 @@ class RecurrentAlgorithm {
   void Run(const framework::Scope& scope,
            const platform::DeviceContext& dev_ctx) const;
 
-  void Init(std::unique_ptr<rnn::Argument> arg) { arg_ = std::move(arg); }
+  void Init(std::unique_ptr<rnn::Argument> arg,
```
Comment (wangkuiyi) [inline paddle/operators/recurrent_op.h]: Why `unique_ptr` here? Could we just do 

```cpp
class RecurrentOp ... {
 private:
  rnn::Argument arg_;
```

File: paddle/operators/recurrent_op.h (hunk 1 above)
Comment (Superjom) [inline paddle/operators/recurrent_op.h]: In `pybind.cc`, a `std::shared_ptr` is passed to `RecurrentOp` as an argument, so here is a shared pointer too.

File: paddle/operators/recurrent_op.h (hunk 2 above)
Comment (Superjom) [inline paddle/operators/recurrent_op.h]: the `rnn::Argument` is a argument that should be passed to `RecurrentAlgorithm`, a unique pointer may be lighter to pass, and `std::move` is used here, finally only `RecurrentAlgorithm` owns this memory.

File: paddle/operators/recurrent_op.h (hunk 1 above)
Comment (wangkuiyi) [inline paddle/operators/recurrent_op.h]: Why do we need shared_ptr in our Python binding?

File: paddle/operators/recurrent_op.h (hunk 2 above)
Comment (wangkuiyi) [inline paddle/operators/recurrent_op.h]: I don't get it. By defining arg_ a data member of RecurrentOp, we can of course call `RecurrentAlgorithm(&arg_);`, could we?

File: paddle/operators/recurrent_op.h (hunk 1 above)
Comment (Superjom) [inline paddle/operators/recurrent_op.h]: just like `NetOp`, RNNOp has its own python binding to expose some method, such as `set_stepnet`.

In `pybind.cc`, all the operators created by user are handled by `shared_ptr`, maybe easier to be referenced across different levels, when user create a `Operator`, a shared_ptr is returned, it is free to use the pointer, or pass the pointer into multiple `NetOp` (add an op into a net).

Operators will live until the whole execution period,  so a `shared_ptr` seems reasonable.

File: paddle/operators/recurrent_op.h (hunk 2 above)
Comment (Superjom) [inline paddle/operators/recurrent_op.h]: get it.

File: paddle/operators/recurrent_op.h (hunk 1 above)
Comment (wangkuiyi) [inline paddle/operators/recurrent_op.h]: Do you mean something like [here](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/pybind.cc#L230):

```cpp
py::class_<operators::NetOp, std::shared_ptr<operators::NetOp>> net(m, "Net");
```

It seem that calling Python `new Net()` returns a `shared_ptr<operator::NetOp>` -- however, when will the reference count become 0 and the C++ instance get freed?

**Pull request merged**